File: python/calc_heavylight.py
import numpy as np 
import itertools as it
from numpy.linalg import inv
import struct as st 
import matplotlib.pyplot as plt 
from csv import writer
import os

def Calc_pseudo(arrayL, arrayH):
    """calc vacuum expectaion value for each time step - pseudoscalar""" 
    pscalar = complex(0,0)
    for x,y,z,c1,c2 in it.product(range(Ns), range(Ns), range(Ns), range(3), range(3)):
        l = np.matrix(arrayL[x,y,z,:2,c1,:2,c2], dtype=complex)
        h = np.matrix(arrayH[x,y,z,:,c1,:,c2], dtype=complex)
        S = l*h.H
        pscalar += np.real(S[0,0] + S[1,1]) 
        
    return pscalar

def Calc_vector(arrayL, arrayH):
    """calc vacuum expectation value for each time step - vector """
    vector_val = 0
    for x,y,z,c1,c2 in it.product(range(Ns), range(Ns), range(Ns), range(3), range(3)):

        l = np.matrix(arrayL[x,y,z,:2,c1,:2,c2])
        h = np.matrix(arrayH[x,y,z,:,c1,:,c2])
        matr_1 = -S15*h*S15*l.H
        vector_val += np.real(matr_1[0,0] + matr_1[1,1])
        matr_2 = -S25*h*S25*l.H
        vector_val += np.real(matr_2[0,0] + matr_2[1,1])
        matr_3 = -S35*h*S35*l.H
        vector_val += np.real(matr_3[0,0] + matr_3[1,1])

    return vector_val

def Calc_correlator(filename_l, filename_h):
 ##import the propagators
    numdoub_l = (Ns**3)*4*4*3*3*2 
    numbyt_l = 8*numdoub_l

    numdoub_h = (Ns**3)*2*2*3*3*2
    numbyt_h = 8*numdoub_h

    with open(filename_l, 'rb') as light, open(filename_h, 'rb') as heavy:#read file and extract prop    
        #save 4 ints at start of file (light)  
        data = light.read(16)
        ints = st.unpack("iiii", data[0:16])
        #save norm
        data = light.read(8)
        norm = st.unpack("d", data[0:8])[0]
        
        #save filesize (heavy)
        data = heavy.read(4)
        size = st.unpack("i", data[0:4])
        
        prop_l = np.zeros((Ns, Ns, Ns, 4, 3, 4, 3), dtype=complex)
        prop_tran = np.zeros((Ns, Ns, Ns, 4, 3, 4, 3), dtype=complex)
        prop_h = np.zeros((Ns, Ns, Ns, 2, 3, 2, 3), dtype=complex)
        pseudo = np.zeros((Nt), dtype=complex)
        vector = np.zeros((Nt), dtype=complex)

        for t in range(Nt):

            data_l = light.read(numbyt_l)
            pointer_l = 0
            
            data_h = heavy.read(numbyt_h)
            pointer_h = 0
            
            #read in light prop
            for x,y,z,s1,c1,s2,c2 in it.product(range(Ns), range(Ns), range(Ns), range(4), range(3), range(4), range(3)):
                prop_l[x,y,z,s1,c1,s2,c2] = (st.unpack("d", data_l[pointer_l:pointer_l+8])[0] + st.unpack("d", data_l[pointer_l+8:pointer_l+16])[0]*1j)

                pointer_l += 16
            
            #change light prop chiral to NR rep
            for x,y,z,c1,c2 in it.product(range(Ns), range(Ns), range(Ns), range(3), range(3)):
                #prop_tran[x,y,z,:,c1,:,c2] = trans_matrix_inv*prop_l[x,y,z,:,c1,:,c2]*trans_matrix*(1/2)
                prop_tran[x,y,z,:,c1,:,c2] = trans_matrix*prop_l[x,y,z,:,c1,:,c2]*trans_matrix_inv


            #read in heavy prop
            for s1,s2,c1,c2,z,y,x in it.product(range(2), range(2), range(3), range(3), range(Ns), range(Ns), range(Ns)):
                prop_h[x,y,z,s1,c1,s2,c2] = (st.unpack("d", data_h[pointer_h:pointer_h+8])[0] + st.unpack("d", data_h[pointer_h+8:pointer_h+16])[0]*1j)

                pointer_h += 16

            pseudo[t] = Calc_pseudo(prop_tran, prop_h)
            vector[t] = Calc_vector(prop_tran, prop_h)

    light.close()
    heavy.close()

    return pseudo, vector

#Define matrices
trans_matrix = np.matrix([[1, 0, -1, 0],
                          [0, 1, 0, -1],
                          [1, 0, 1, 0],
                          [0, 1, 0, 1]], dtype = complex)
trans_matrix_inv = np.linalg.inv(trans_matrix)

Sigma0 = np.matrix([[1,0],
                    [0,1]], dtype=complex)
Sigma1 = np.matrix([[0,1],
                    [1,0]], dtype=complex)
Sigma2 = np.matrix([[0,-1j],
                    [1j,0]], dtype=complex)
Sigma3 = np.matrix([[1,0],
                    [0,-1]], dtype=complex)
Sigma5 = Sigma1*Sigma2*Sigma3*Sigma0
S15 = Sigma1*Sigma5
S25 = Sigma2*Sigma5
S35 = Sigma3*Sigma5

Gamma0 = np.matrix([[1, 0, 0, 0],
                    [0, 1, 0, 0],
                    [0, 0, -1, 0],
                    [0, 0, 0, -1]], dtype=complex)
Gamma1 = np.matrix([[0, 0, 0, -1j],
                    [0, 0, -1j, 0],
                    [0, 1j, 0, 0],
                    [1j, 0, 0, 0]], dtype=complex)
Gamma2 = np.matrix([[0, 0, 0, -1],
                    [0, 0, 1, 0],
                    [0, 1, 0, 0],
                    [-1, 0, 0, 0]], dtype=complex)
Gamma3 = np.matrix([[0, 0, -1j, 0],
                    [0, 0, 0, 1j],
                    [1j, 0, 0, 0],
                    [0, -1j, 0, 0]], dtype=complex)
Gamma5 = Gamma1*Gamma2*Gamma3*Gamma0
G15 = Gamma1*Gamma5
G25 = Gamma2*Gamma5
G35 = Gamma3*Gamma5


######################################################################################

Ns=32
Nt=32

#light_dir = f'/home/postgrad/rachelhd/Bmeson/rqcd/propagators/{Ns}x{Nt}/strange/prop'
light_dir = f'/data/rachel/Bmeson/rqcd/propagators/{Ns}x{Nt}/s0'
heavy_dir = f'/data/rachel/Bmeson/nrqcd/propagators/{Ns}x{Nt}/prop/sprop'
file_dir =  f'/home/postgrad/rachelhd/Bmeson/hl_corrs/Gen2'

l_files = os.listdir(light_dir)    #list files in directory
h_files = os.listdir(heavy_dir)
l_files.sort()                     #sort because python doesn't
h_files.sort()
#print(l_files, h_files)
if len(l_files) > len(h_files):
    number_of_cfgs = len(l_files)
else:
    number_of_cfgs = len(h_files)
print(number_of_cfgs)
pseudo = []
vector = []

numcfgs=0
for i in range(6010,6020,10):
    if any(str(i) in x for x in l_files) and any(str(i) in x for x in h_files):
        ps, v = Calc_correlator(light_dir+f'/Gen2_{Ns}x{Nt}cfgn{i}.s0.m0',  heavy_dir+f'/sprop.{Ns}x{Nt}_{i}')
        cfg = np.array([i])
        ps = np.hstack((cfg, ps))
        v = np.hstack((cfg, v))
        pseudo.append(np.real(ps))
        vector.append(np.real(v))
        with open(file_dir+f'/hl_ps_{Ns}x{Nt}.csv', 'a') as file:
            write = writer(file)
            write.writerow(pseudo[numcfgs])
            file.close()

        with open(file_dir+f'/hl_v_{Ns}x{Nt}.csv', 'a') as file:
            write = writer(file)
            write.writerow(vector[numcfgs])
            file.close()
        numcfgs+=1
    else:
        print(f'no {i}')
<<<<<<< HEAD






=======
>>>>>>> c

File: C/calc_heavylight.c
#include <stdio.h>
#include <stdlib.h>
#include <complex.h>
#include <stdint.h>
#include <errno.h>

#define Ns 32 
#define Nt 28
#define START 4000
#define STOP 4100
#define STEP 10

// index the flat prop_l/h array
size_t idx_light(int x, int y, int z, int s1, int c1, int s2, int c2, int t) {
    return (((((((size_t)x * Ns + y) * Ns + z) * 4 + s1) * 3 + c1) * 4 + s2) * 3 + c2) *Nt+ t;
}
size_t idx_heavy(int x, int y, int z, int s1, int c1, int s2, int c2, int t) {
    return (((((((size_t)x * Ns + y) * Ns + z) * 2 + s1) * 3 + c1) * 2 + s2) * 3 + c2) *Nt+t;
}

void mat4_mult(const double complex A[4][4], const double complex B[4][4], double complex C[4][4]) {
    for (int i = 0; i < 4; i++)
    	for (int j = 0; j < 4; j++) {
            C[i][j] = 0.0;
            for (int k = 0; k < 4; k++){
          	C[i][j] += A[i][k] * B[k][j];
      	    }
	}
    }

// transform light to NR rep
void transform_light(
    const double complex *prop_l,  // input, chiral basis
    double complex *prop_tr,     // output, NR basis
    const double complex trans[4][4], //transform mat
    const double complex trans_inv[4][4] 
) {
    for (int t=0;t<Nt;t++){	
    for (int x = 0; x < Ns; x++){
    	for (int y = 0; y < Ns; y++){
    	    for (int z = 0; z < Ns; z++){
    		for (int c1 = 0; c1 < 3; c1++){
    		    for (int c2 = 0; c2 < 3; c2++){
			// Extract 4x4 block for fixed x,y,z,c1,c2
			double complex block[4][4];
			for (int s1 = 0; s1 < 4; s1++){
			    for (int s2 = 0; s2 < 4; s2++){
			    	size_t idx = idx_light(x, y, z, s1, c1, s2, c2, t);
			    	block[s1][s2] = prop_l[idx];
			    }
			}
			// temp1 = trans * block
			double complex temp1[4][4];
			mat4_mult(trans, block, temp1);

			// temp2 = temp1 * trans_inv
			double complex temp2[4][4];
			mat4_mult(temp1, trans_inv, temp2);

			// Store result back, scaled by 1/2
			for (int s1 = 0; s1 < 4; s1++){
			    for (int s2 = 0; s2 < 4; s2++){
			    	size_t idx = idx_light(x, y, z, s1, c1, s2, c2, t);
			    	prop_tr[idx] = temp2[s1][s2] * 0.5;
		            }
			}
		    }
		}
	    }
	}
    }
    }
}

//read binary light prop file
void read_light_prop(const char *filename, 
                     double complex *prop_l){
    FILE *f = fopen(filename, "rb");
    if (!f){
        //fprintf(stderr, "Unable to open file '%s': %d\n", filename, strerror(errno));
        perror("Error opening file");
	exit(1);
    }
    
    // Read header
    int header[4];
    double norm;
    if (fread(header,sizeof(int),4,f)!=4 ||
    	fread(&norm,sizeof(double),1,f)!=1) {
    	fclose(f);
	fprintf(stderr, "Error reading header/norm from file %s\n", filename);
	exit(1);	
    }

    //Read prop data
    for (int t=0; t<Nt; t++){
    for (int x=0; x<Ns; x++){
  	for (int y=0; y<Ns; y++){
    	    for (int z=0; z<Ns; z++){
    		for (int s1=0; s1<4; s1++){
		    for (int c1=0; c1<3; c1++){
    			for (int s2=0; s2<4; s2++){
			    for (int c2=0; c2<3; c2++){
    				double repart, impart;
				if (fread(&repart,sizeof(double),1,f)!=1 ||
				    fread(&impart,sizeof(double),1,f)!=1){
				    fprintf(stderr, "Error reading data at t=%d x=%d y=%d z=%d s1=%d c1=%d s2=%d c2=%d\n",
                                            t, x, y, z, s1, c1, s2, c2);
				    fclose(f);
				    exit(1);
				}
				size_t idx = idx_light(x, y, z, s1, c1, s2, c2, t); 
				prop_l[idx] = repart+impart*I;
    			    }
    			}
		    }
		}
	    }
	}
    }
    }
    fclose(f);
}

//read binary heavy prop file
void read_heavy_prop(const char *filename,
                     double complex *prop_h){
    FILE *f = fopen(filename, "rb");
    if (!f){
        //fprintf(stderr, "Unable to open file '%s': %d\n", filename, strerror(errno));
        perror("Error opening file");
	exit(1);
        }

    // Read header
    int header;
    if (fread(&header,sizeof(int),1,f) !=1){
	perror("Error reading header from heavy file");
	fclose(f);
	exit(1);	
    } 

    //Read prop data
    for (int t=0; t<Nt; t++){
    for (int s1=0; s1<2; s1++){
        for (int s2=0; s2<2; s2++){
            for (int c1=0; c1<3; c1++){
                for (int c2=0; c2<3; c2++){
                    for (int z=0; z<Ns; z++){
                        for (int y=0; y<Ns; y++){
                            for (int x=0; x<Ns; x++){
                                double repart, impart;
                                if (fread(&repart,sizeof(double),1,f)!=1 ||
                                    fread(&impart,sizeof(double),1,f)!=1){
				    fprintf(stderr, "Error reading data at t=%d x=%d y=%d z=%d s1=%d c1=%d s2=%d c2=%d\n",
                                           t, x, y, z, s1, c1, s2, c2);
                                    fclose(f);
                                    exit(1);
                            	}
				size_t idx = idx_heavy(x, y, z, s1, c1, s2, c2, t);
				prop_h[idx] = repart + impart*I;
			    }
                        }
                    }
                }
            }
        }
    }
    }
    fclose(f);
}

void calc_pseudo(const double complex *arrayL, const double complex *arrayH, double *pseudo){
    for (int t = 0; t < Nt; t++){
	double pscalar = 0.0;
	for (int x = 0; x < Ns; x++){
	    for (int y = 0; y < Ns; y++){
		for (int z = 0; z < Ns; z++){
		    for (int c1 = 0; c1 < 3; c1++){
			for (int c2 = 0; c2 < 3; c2++){
			// light: [x][y][z][0..1][c1][0..1][c2]
		        // heavy: [x][y][z][0..1][c1][0..1][c2]

	         	double complex l00 = arrayL[idx_light(x, y, z, 0, c1, 0, c2, t)];
			double complex l01 = arrayL[idx_light(x, y, z, 0, c1, 1, c2, t)];
			double complex l10 = arrayL[idx_light(x, y, z, 1, c1, 0, c2, t)];
			double complex l11 = arrayL[idx_light(x, y, z, 1, c1, 1, c2, t)];

			double complex h00 = arrayH[idx_heavy(x, y, z, 0, c1, 0, c2, t)];
			double complex h01 = arrayH[idx_heavy(x, y, z, 0, c1, 1, c2, t)];
			double complex h10 = arrayH[idx_heavy(x, y, z, 1, c1, 0, c2, t)];
			double complex h11 = arrayH[idx_heavy(x, y, z, 1, c1, 1, c2, t)];

			// Matrix product and Hermitian conjugate (transpose + conjugate)
			// S = l * h.H, where l and h are 2x2 matrices
			// S[0,0] = l00*conj(h00) + l01*conj(h01)
			// S[1,1] = l10*conj(h10) + l11*conj(h11)
			double complex S00 = l00*conj(h00) + l01*conj(h01);
			double complex S11 = l10*conj(h10) + l11*conj(h11);

			pscalar += creal(S00 + S11); //trace of S
			}
		    }
		}
            }
	}
	pseudo[t] = pscalar;
    }
}

void calc_vector(const double complex *arrayL, const double complex *arrayH, double *vector){
    
    double complex S15[2][2]={
	    {0.0 + 0.0*I, 0.0 + 1.0*I},
	    {0.0 + 1.0*I, 0.0 + 0.0*I}
    };
    double complex S25[2][2]={
            {0.0 + 0.0*I, 1.0 + 0.0*I},
            {-1.0 + 0.0*I, 0.0 + 0.0*I}
    };
    double complex S35[2][2]={
            {0.0 + 1.0*I, 0.0 + 0.0*I},
            {0.0 + 0.0*I, 0.0 - 1.0*I}
    };
    for (int t = 0; t < Nt; t++){
        double vectorval = 0.0;
        for (int x = 0; x < Ns; x++){
            for (int y = 0; y < Ns; y++){
                for (int z = 0; z < Ns; z++){
                    for (int c1 = 0; c1 < 3; c1++){
                        for (int c2 = 0; c2 < 3; c2++){
                        // light: [x][y][z][0..1][c1][0..1][c2]
                        // heavy: [x][y][z][0..1][c1][0..1][c2]

                        double complex l00 = arrayL[idx_light(x, y, z, 0, c1, 0, c2, t)];
                        double complex l01 = arrayL[idx_light(x, y, z, 0, c1, 1, c2, t)];
                        double complex l10 = arrayL[idx_light(x, y, z, 1, c1, 0, c2, t)];
                        double complex l11 = arrayL[idx_light(x, y, z, 1, c1, 1, c2, t)];

                        double complex h00 = arrayH[idx_heavy(x, y, z, 0, c1, 0, c2, t)];
                        double complex h01 = arrayH[idx_heavy(x, y, z, 0, c1, 1, c2, t)];
                        double complex h10 = arrayH[idx_heavy(x, y, z, 1, c1, 0, c2, t)];
                        double complex h11 = arrayH[idx_heavy(x, y, z, 1, c1, 1, c2, t)];

                        // Matrix product and Hermitian conjugate (transpose + conjugate)
                        // Ti = Si5.H * h * Si5 l.H, where l and h are 2x2 matrices
                        // T1/2[0,0] =  
                        double complex T100 = -S15[0][1]*h10*S15[0][1]*conj(l01) -S15[0][1]*h11*S15[1][0]*l00;
			double complex T111 = -S15[1][0]*h00*S15[0][1]*l11 -S15[1][0]*h00*S15[1][0]*conj(l10);
                        vectorval += creal(T100 + T111); //trace of T1
                        
			double complex T200 = -S25[0][1]*h10*S25[0][1]*conj(l01) -S25[0][1]*h11*S25[1][0]*l00;
                        double complex T211 = -S25[1][0]*h00*S25[0][1]*l11 -S25[1][0]*h00*S25[1][0]*conj(l10);
                        vectorval += creal(T200 + T211); //trace of T1
			
			double complex T300 = -S35[0][0]*h00*S35[0][0]*l00 -S35[0][0]*h01*S35[1][1]*conj(l01);
                        double complex T311 = -S35[1][1]*h10*S35[0][0]*conj(l10) -S35[1][1]*h11*S35[1][1]*l11;
                        vectorval += creal(T300 + T311); //trace of T1

			}
                    }
                }
            }
        }
        vector[t] = vectorval;
    }

}
//////////////////////////////////////////////////////////////////////
int main(){

double pseudo_sum[Nt];
double pseudo_avg[Nt];
double vector_sum[Nt];

//double pseudo[Nt];
//double vector[Nt];
//for (int i=0; i<Nt; i++){
//    pseudo[i] = 0.0;
//    vector[i] = 0.0;
//}
double complex trans[4][4] = {
    {1.0 + 0.0*I, 0.0 + 0.0*I, -1.0 + 0.0*I, 0.0 + 0.0*I},
    {0.0 + 0.0*I, 1.0 + 0.0*I,  0.0 + 0.0*I, -1.0 + 0.0*I},
    {1.0 + 0.0*I, 0.0 + 0.0*I,  1.0 + 0.0*I,  0.0 + 0.0*I},
    {0.0 + 0.0*I, 1.0 + 0.0*I,  0.0 + 0.0*I,  1.0 + 0.0*I}
};
double complex trans_inv[4][4];
for (int i = 0; i < 4; i++){
    for (int j = 0; j < 4; j++){
        trans_inv[i][j] = conj(trans[j][i]);
    }
}
//for (int nt=0; nt<Nt; nt++){
    double complex *prop_l = malloc(sizeof(double complex)*Ns*Ns*Ns*4*3*4*3*Nt);
    double complex *prop_h = malloc(sizeof(double complex)*Ns*Ns*Ns*2*3*2*3*Nt);
    double complex *prop_tr = malloc(sizeof(double complex)*Ns*Ns*Ns*4*3*4*3*Nt);
    if (!prop_l || !prop_h || !prop_tr) {
    	perror("malloc prop_l/h/tr");
	free(prop_l);
	free(prop_h);
	free(prop_tr);
    	exit(1);
    }
    
    char hfile[256];
    char lfile[256];
    int ncfg = 0;
    for (int cfg=START; cfg<=STOP; cfg+=STEP){
    	double pseudo[Nt];
    	double vector[Nt];
    	for (int i=0; i<Nt; i++){
            pseudo[i] = 0.0;
            vector[i] = 0.0;
    	}
 
    	sprintf(hfile,"/data/rachel/Bmeson/nrqcd/propagators/%dx%d/prop/sprop/sprop.%dx%d_%d", 
		    Ns,Nt,Ns,Nt,cfg);
    	sprintf(lfile, "/data/rachel/Bmeson/rqcd/propagators/%dx%d/light/s0/Gen2_%dx%dcfgn%d.s0.m0", 
		    Ns,Nt,Ns,Nt,cfg);
	
	// Check heavy file
        FILE *fh = fopen(hfile, "rb");
        if (!fh) {
            printf("Warning: Cannot open heavy file for config %d, skipping.\n", cfg);
            continue;
        }
        fclose(fh);

        // Check light file
        FILE *fl = fopen(lfile, "rb");
        if (!fl) {
            printf("Warning: Cannot open light file for config %d, skipping.\n", cfg);
            continue;
        }
        fclose(fl);

    	read_heavy_prop(hfile, prop_h);
    	read_light_prop(lfile, prop_l);
    	printf("Files read successfully!\n");
    
    	transform_light(prop_l, prop_tr, trans, trans_inv);
    	calc_pseudo(prop_tr, prop_h, pseudo);
 	printf("Calculated pseudoscalar:");
    	for (int t = 0; t < Nt; t++) {
	    printf("pseudo[%d] = %f\n", t, pseudo[t]);
    	    pseudo_sum[t] += pseudo[t];
	}
	calc_vector(prop_tr, prop_h, vector);
	printf("Calculated vector:");
	for (int t = 0; t < Nt; t++) {
            printf("vector[%d] = %f\n", t, vector[t]);
            vector_sum[t] += vector[t];
        }
	ncfg++;
    }
    
    for (int t=0;t <=Nt; t++){
	pseudo_avg[t] = pseudo_sum[t] / ncfg;
    }
    printf("\nAveraged pseudoscalar over %d configs:\n", ncfg);
    for (int t = 0; t < Nt; t++) {
        printf("pseudo_avg[%d] = %f\n", t, pseudo_avg[t]);
    }

    free(prop_l);
    free(prop_h);
    free(prop_tr);


    return 0;

}

